### PR TITLE
Fix analysis time

### PR DESCRIPTION
**Pull request by yi-fan-wang**
Try to relax the constraint that data from all detectors should have the same length. Work in progress

2021.11.08 Update: This module adds support to `gaussian_noise`, `marginalized_phase` and `marginalized_polarization` for analyzing different data segment lengths and/or sampling rates for different IFOs.

Future optimization: add support for more models, handle the case when some of the IFOs use the same settings (in that case I don't need to generate a waveform generator for each one of IFO

### CODE REVIEW

**Review by ahnitz — CHANGES_REQUESTED** [review]
@yi-fan-wang On first glance this seems largely OK. 

One thing to fix is that this doesn't affect the marginalized_phase model at the moment. I think the easiest way to do that would actually be to have the marginalized_phase class inherit from the gaussianNoise (rather than basegaussiannoise) and have it *not* redo the __init__. Does that make sense to you and if you look at what both __init__'s are doing, do you agree that would work as intended? The marginalized polarization needs a different init because it has to use a different waveform generator due to needing both polarizations returned, however. 

For future reference, this could be sped up slightly in the case where some of the detectors use the same setting, but we can leave that for a future PR.

File: pycbc/inference/models/marginalized_gaussian_noise.py [hunk 1]
```diff
@@ -262,7 +274,12 @@ def _loglr(self):
         """
         params = self.current_params
         try:
-            wfs = self.waveform_generator.generate(**params)
+            if self.all_ifodata_same_rate_length:
```
Comment (ahnitz) [inline pycbc/inference/models/marginalized_gaussian_noise.py:277]: Marginalized phase might also need this bit.

**Comment by ahnitz** [comment]
@yi-fan-wang Besides for my comment, do your tests show that this is working as expected? If so, we should try to address the final issues and get this merged.

@sum33it Any other comments?

**Review by ahnitz — APPROVED** [review]
@yi-fan-wang Please merge if you are satisfied this is correct (i.e. works in your tests) and the unittests pass

**Comment by yi-fan-wang** [comment]
@ahnitz I'll do another PE run to make sure the `marginalized_phase` is working as well and then I'll merge it.

File: pycbc/inference/models/marginalized_gaussian_noise.py [hunk 2]
```diff
@@ -28,7 +28,7 @@
 from .gaussian_noise import (BaseGaussianNoise, create_waveform_generator)
```
Comment (ahnitz) [inline pycbc/inference/models/marginalized_gaussian_noise.py:28]: @yi-fan-wang You forgot to import GaussianNoise when you made this change

**Comment by ahnitz** [comment]
@yi-fan-wang Please check this over and fix the issues so this can be merged.

File: pycbc/inference/models/marginalized_gaussian_noise.py (hunk 1 above)
Comment (yi-fan-wang) [reply pycbc/inference/models/marginalized_gaussian_noise.py:277]: Added for marginalized phase

File: pycbc/inference/models/marginalized_gaussian_noise.py (hunk 2 above)
Comment (yi-fan-wang) [reply pycbc/inference/models/marginalized_gaussian_noise.py]: Fixed

**Pull request merged**